lcls-twincat-motion/Library/POUs/Motion/States/FB_StateSetupHelper.TcPOU

```diff
@@ -0,0 +1,107 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1">
+  <POU Name="FB_StateSetupHelper" Id="{802f49f0-571f-413b-bf0f-1cf473b320d9}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_StateSetupHelper
+(*
+    This is a helper for setting up large numbers of ST_PositionState instances.
+
+    This is typically verbose to do by hand in the normal ways and can be error-prone.
+
+    Calling with bSetDefault:=TRUE will set the default values to all the values
+    from the input stPositionState. Note that the other args will be ignored in this case.
+
+    Calling without bSetDefault or with it set to FALSE will apply values to the
+    input stPositionState with the following priority order:
+
+    1. The value used in the function block call
+    2. The value from the template stPositionState used in the most recent
+       call with bSetDefault:=TRUE
+    3. The original default value as defined on ST_PositionState
+
+    Example expected usage:
+    VAR
+        fbStateSetup: FB_StateSetupHelper;
+        stDefault: ST_PositionState := (
+            fDelta := 0.5,
+            fVelocity := 10,
+            bMoveOk := TRUE,
+            bValid := TRUE
+        );
+        astStates1: ARRAY[1..GeneralConstants.MAX_STATES] OF ST_PositionState;
+        astStates2: ARRAY[1..GeneralConstants.MAX_STATES] OF ST_PositionState;
+        astStates2: ARRAY[1..GeneralConstants.MAX_STATES] OF ST_PositionState;
+    END_VAR
+
+    fbStateSetup(stPositionState:=stDefault, bSetDefault:=TRUE);
+
+    fbStateSetup(stPositionState:=astStates1[1], sName:='OUT', fPosition:=10);
+    fbStateSetup(stPositionState:=astStates1[2], sName:='YAG', fPosition:=20);
+    fbStateSetup(stPositionState:=astStates1[3], sName:='TT', fPosition:=30);
+
+    fbStateSetup(stPositionState:=astStates2[1], sName:='OUT', fPosition:=-30);
+    fbStateSetup(stPositionState:=astStates2[2], sName:='YAG', fPosition:=35);
+    fbStateSetup(stPositionState:=astStates2[3], sName:='TT', fPosition:=70);
+
+    fbStateSetup(stPositionState:=astStates3[1], sName:='OUT', fPosition:=0.4, fDelta:=0.1);
+    fbStateSetup(stPositionState:=astStates3[2], sName:='YAG', fPosition:=2.3, fDelta:=0.1);
+    fbStateSetup(stPositionState:=astStates3[3], sName:='TT', fPosition:=5.6, fDelta:=0.1;
+*)
```

Comment:
Is it intended that this is only called during initialization/startup? I think that the example, as written, would continually over-write the values, preventing any modifications from EPICS. Is this something we want to prevent? 

Regardless, the code looks good. I like it.

Reply:
You're right, the example as-written will implicitly block EPICS from making further changes. I think it makes sense to adjust the examples and make a note of why everything is wrapped inside a "call me only once" block.

Reply:
I guess one more follow-up thought is: is there any possible way for the function block itself to exercise a "do the thing only once" sort of behavior? It's worth a short brainstorm I think, since bInit only-run-me-once blocks are at least a bit annoying to write. Time to ponder...

Reply:
A simple adjustment to the function block code is to skip overwriting structs that already have writes to either the sName or bValid fields

Reply:
I tried to resolve this with https://github.com/pcdshub/lcls-twincat-motion/pull/184/commits/809f126477d7c7a9b5cdc90d93dcdb2303a95693

Reply:
I think that's a reasonable tweak, Zach.

I'm personally somewhat in favor of the `bInit` mechanism (at least partly because I can reset `bInit` and make it redo the initialization steps whenever I choose), but I understand the desire to use it as you mention @slactjohnson . This seems a reasonable middle-ground for both use cases.

Reply:
@klauer I don't really favor one way or the other; I see advantages to both. But in principle you could toggle `bForceUpdate`, right? And that way, you only re-init one thing.

Reply:
`bForceUpdate` is per function block call rather than implicitly for all of them through an `IF bInit THEN` block. 

If you tied a global `bReinitialize` variable to all `bForceUpdate` inputs (or once as like `fbSetDefaults.bForceUpdate := bReinitialize;`) it'd even out in the end, I think

Reply:
I think the reason I went with this route instead of doing a docs revision is because calling these multiple times will break mirror-like state blocks that set their positions based on encoder counts. If there wasn't a potential PLC bug associated with running this twice I think the explicit bInit version is preferable.

The without this, the flow for the mirror-likes would be something like:

- set fPosition to 0
- set nEncoderCount to <some number>
- states FB calculates corrected fPosition value, and sets it
- various elements use the calculated fPosition
- next cycle: set fPosition to 0 (maybe even _during_ a move!)

So I'd like to avoid this kind of behavior altogether

Reply:
I think actually bForceUpdate runs on all of them for one FB since it doesn't get reset anywhere in the function block call